python/python/bystro/ancestry/train.py

```diff
@@ -308,6 +313,88 @@ def _calc_fst(variant_counts: pd.Series, samples: pd.DataFrame) -> float:
+    #If testing other vcfs this line will also need to be altered
+    dosage_vcf=dosage_vcf.rename(
+        columns={"#CHROM": "Chromosome", "POS": "Position"}, 
+        errors="raise"
+    )
+    return dosage_vcf
+    
+
+def load_pca_loadings(GNOMAD_PC_PATH: str, dosage_vcf: pd.DataFrame) -> pd.DataFrame:
+    #Gnomad pc file 
+    """Load in the gnomad PCs and reformat for PC transformation.
+    """
+    loadings = pd.read_csv(GNOMAD_PC_PATH, sep="\t")
+    #Gnomad pc loadings file includes additional formatting that needs to be sanitized 
+    loadings[["Chromosome", "Position"]] = loadings["locus"].str.split(":", expand=True)
+    #Match variant format of gnomad loadings with 1kgp vcf
+    get_chr_pos = lambda x: x["locus"][3:]
+    get_ref_allele = lambda x: x["alleles"][2]
+    get_alt_allele = lambda x: x["alleles"][6]
+    loadings["variant"] = loadings.apply(
+        lambda x: ":".join([get_chr_pos(x), get_ref_allele(x), get_alt_allele(x)]), axis=1
+    )
+    #Remove brackets
+    loadings["loadings"] = loadings["loadings"].str[1:-1]
+    #Split PCs and join back
+    gnomadPCs = loadings["loadings"].str.split(",", expand=True)
+    loadings = loadings.join(gnomadPCs)
+    loadings = loadings.reset_index()
+    pc_range = range(8, 38)
+    pc_loadings = loadings.iloc[:, pc_range].copy()
+    pc_loadings["variant"] = loadings["variant"]
+    pc_loadings = pc_loadings.set_index("variant")
+    pc_loadings = pc_loadings.sort_index()
+    return pc_loadings
+
+
+def loadings_var_overlap(pc_loadings: pd.DataFrame, dosage_vcf: pd.DataFrame) -> tuple[pd.DataFrame, pd.DataFrame]:
+    """Check overlap between gnomad loadings and reference vcf"""
+    #Remove variants that are missing from IGSR version of 1kgp
+    loadings_var_set=set(pc_loadings.index)
+    dosage_var_set=set(dosage_vcf.ID)
+    missing_variants = loadings_var_set - dosage_var_set
+    pc_loadings_overlap=pc_loadings[~pc_loadings.index.isin(missing_variants)]
+    var_overlap = loadings_var_set.intersection(dosage_var_set)
+    assert len(pc_loadings_overlap) == len(var_overlap)
+    assert len(dosage_vcf) == len(var_overlap)
+    return pc_loadings_overlap
+
+
+def apply_pca_transform(pc_loadings_overlap: pd.DataFrame, dosage_vcf: pd.DataFrame) -> pd.DataFrame:
+    """Transform vcf with genotypes in dosage format with PCs loadings from gnomad PCA."""
+    #Only genos are required from dosage_vcf for transformation step
+    genos = dosage_vcf.iloc[:, 9:]
+    genos["variant"] = dosage_vcf["ID"]
+    genos=genos.set_index("variant")
+    genos=genos.sort_index()
+    #Transpose genos so it is in the correct configuration for dot product
+    genos_transpose=genos.T
+    #Ensure that genos_transpose and pc_loadings have the same variants in same order
+    assert (genos_transpose.columns == pc_loadings_overlap.index).all()
+    assert genos_transpose.shape[1] == pc_loadings_overlap.shape[0]
+    #Dot product
+    transformed_data = genos_transpose @ pc_loadings_overlap
```

Comment:
if we preprocess the dfs a bit more upfront, the whole function could be this line, right?